Tests/AmplitudeTests/AmplitudeSessionTests.swift

```diff
@@ -0,0 +1,355 @@
+    }
+
+    func testDistantForegroundBackgroundEventsShouldStartNewSession() throws {
+        let lastEventId: Int64 = 123
+        try storageMem.write(key: StorageKey.LAST_EVENT_ID, value: lastEventId)
+
+        let amplitude = Amplitude(configuration: configuration)
+
+        let eventCollector = EventCollectorPlugin()
+        amplitude.add(plugin: eventCollector)
+
+        amplitude.onEnterForeground(timestamp: 1000)
+        amplitude.track(event: BaseEvent(userId: "user", timestamp: 1500, eventType: "test event 1"))
+        amplitude.onExitForeground(timestamp: 2000)
+        amplitude.track(event: BaseEvent(userId: "user", timestamp: 3000, eventType: "test event 2"))
+
+        let collectedEvents = eventCollector.events
+
+        XCTAssertEqual(collectedEvents.count, 5)
+
+        var event = collectedEvents[0]
+        XCTAssertEqual(event.eventType, Constants.AMP_SESSION_START_EVENT)
+        XCTAssertEqual(event.sessionId, 1000)
+        XCTAssertEqual(event.timestamp, 1000)
+        XCTAssertEqual(event.eventId, lastEventId+1)
+
+        event = collectedEvents[1]
+        XCTAssertEqual(event.eventType, "test event 1")
+        XCTAssertEqual(event.sessionId, 1000)
+        XCTAssertEqual(event.timestamp, 1500)
+        XCTAssertEqual(event.eventId, lastEventId+2)
+
+        event = collectedEvents[2]
+        XCTAssertEqual(event.eventType, Constants.AMP_SESSION_END_EVENT)
+        XCTAssertEqual(event.sessionId, 1000)
+        XCTAssertEqual(event.timestamp, 2000)
+        XCTAssertEqual(event.eventId, lastEventId+3)
+
+        event = collectedEvents[3]
+        XCTAssertEqual(event.eventType, Constants.AMP_SESSION_START_EVENT)
+        XCTAssertEqual(event.sessionId, 3000)
+        XCTAssertEqual(event.timestamp, 3000)
+        XCTAssertEqual(event.eventId, lastEventId+4)
+
+        event = collectedEvents[4]
+        XCTAssertEqual(event.eventType, "test event 2")
+        XCTAssertEqual(event.sessionId, 3000)
+        XCTAssertEqual(event.timestamp, 3000)
+        XCTAssertEqual(event.eventId, lastEventId+5)
+    }
+
+    func testSessionDataShouldBePersisted() throws {
+        let amplitude1 = Amplitude(configuration: configuration)
+        amplitude1.onEnterForeground(timestamp: 1000)
+
+        XCTAssertEqual(amplitude1.sessionId, 1000)
+        XCTAssertEqual(amplitude1.sessions.sessionId, 1000)
+        XCTAssertEqual(amplitude1.sessions.lastEventTime, 1000)
+        XCTAssertEqual(amplitude1.sessions.lastEventId, 1)
+
+        amplitude1.track(event: BaseEvent(userId: "user", timestamp: 1200, eventType: "test event 1"))
+
+        XCTAssertEqual(amplitude1.sessionId, 1000)
+        XCTAssertEqual(amplitude1.sessions.sessionId, 1000)
+        XCTAssertEqual(amplitude1.sessions.lastEventTime, 1200)
+        XCTAssertEqual(amplitude1.sessions.lastEventId, 2)
+
+        let amplitude2 = Amplitude(configuration: configuration)
+
+        XCTAssertEqual(amplitude2.sessionId, 1000)
```

Comment:
Why different amplitude instances share the same session id?

Reply:
The test checks that new Amplitude instance with the same storage restores sessionId, lastEventTime and lastEventId from the storage and does not start from scratch.